#164069216 Create descriptive signup/registration validation error messages

wasibani-roy opened the pull request:
**What does this PR do?**
This PR creates validation/error messages to give users descriptive feedback on any errors encountered or success archived 

**Description of the tasks to be completed?**
- Create validation messages to handle email input 
- Create validation messages to handle password input 
- Validate that the email used has not been used before to create an account
- Validate that the username chosen is not already in use by another person
- write tests to ensure that the validations are working

**How should this be manually tested?**

- clone this repository 
- cd into Ah-backend-aquaman and checkout to branch ft-validate-registration-data-164069216
- set up your virtual environment
- Then install the requirements pip install -r requirements
- Next run following commands 
     * pip install -r requirements.txt
     * python manage.py makemmigrations
     * python manage.py migrate
     * python manage.py runserver
- Navigate to the /api/users/ route and attempt to enter invalid data

**What are the relevant pivotal tracker stories?**
   * [#164069216](https://www.pivotaltracker.com/story/show/164069216)

**Screenshots**
<img width="1408" alt="screenshot 2019-02-28 at 13 01 17" src="https://user-images.githubusercontent.com/40163337/53558668-87eb8c80-3b59-11e9-8add-e1965b59d47c.png">

CryceTruly commented on authors/apps/authentication/serializers.py:
```diff
@@ -13,7 +15,10 @@ class RegistrationSerializer(serializers.ModelSerializer):
         min_length=8,
-        write_only=True
+        write_only=True,
+        error_messages={
```
@wasibani-roy , since the minimum password length is 8 characters,we could have a message like 
> password should be at least 8 characters long

wasibani-roy replied:
Okay worked on it

CryceTruly commented on authors/apps/authentication/serializers.py:
```diff
@@ -13,7 +15,10 @@ class RegistrationSerializer(serializers.ModelSerializer):
-        write_only=True
+        write_only=True,
+        error_messages={
+            "min_length": "Password must be more than {min_length} characters"
```
something like ``` "min_length": "Password should be atleast  {min_length} characters"```

wasibani-roy replied:
I think this is the same as the one above

CryceTruly commented on authors/apps/authentication/serializers.py:
```diff
@@ -25,6 +30,51 @@ class Meta:
+
+        if user_query_set.exists():
+            raise serializers.ValidationError(
+                "This email has already been used to create a user")
```
hey @wasibani-roy, from a novice user`s perspective, I think ``` "This email has already been used to create a user"``` is confusing, i think it could be better when stated like  
*  "This email has already been used to create an account" or
* "This email is already in use" or 
*  "This email is taken"

wasibani-roy replied:
Changed this thanks

CryceTruly commented on authors/apps/authentication/serializers.py:
```diff
@@ -25,6 +30,51 @@ class Meta:
+        # validate that username passed is correct
+        if not re.search(username_regex, username):
+            raise serializers.ValidationError(
+                "The username is incorrect please use characters")
```
hey @wasibani-roy, from the error message ``` "The username is incorrect please use characters"``` a character can mean anything (a letter,a number,punctualtion marks etc) [Definition of a character](https://en.wikipedia.org/wiki/Character_(computing)")
from the regex check i think we could write the error as 
> The username is invalid, it should contain only letters,

wasibani-roy replied:
Still not very sure about this please throw more light

CryceTruly commented:
LGTM

CryceTruly reviewed (approved):
LGTM

CryceTruly reviewed (dismissed):
LGTM

BagzieGracious reviewed (dismissed):
LGTM

Peace-Apple reviewed (dismissed):
LGTM

Commit:
feat(Registration): Validate user registration data
- Ensure that email submitted by user is correct
- Ensure that password is more than 8 characters and alphanumeric
- Ensure the the email used has not been used to create a previous account
- Ensure that the username is not already in use within the system
[Finishes #164069216]

BagzieGracious reviewed (commented):
good work, but please can you squash 3 commits to 1

Peace-Apple commented:
Good work Roy, however work on squashing the commits

BagzieGracious commented on authors/apps/authentication/tests/test_registration.py:
```diff
@@ -49,3 +42,27 @@ def test_email_exits(self):
         self.assertEqual(
             response.data['errors']['email'][0], "user with this email already exists.")
+
+    def test_register_invalid_email(self):
```
Thanks for the hard, adding other tests and raising the test coverage. but please can you add doc strings for more clarity on your tests.

CryceTruly reviewed (dismissed):
LGTM

BagzieGracious reviewed (dismissed):
good work @wasibani-roy, i have looked through it and its really working

CryceTruly reviewed (dismissed):
LGTM

Peace-Apple reviewed (dismissed):
Good work Roy

CryceTruly reviewed (approved):
LGTM

Peace-Apple reviewed (approved):
LGTM

BagzieGracious reviewed (approved):
@wasibani-roy, i have reviewed your work, you have implemented all the feedback. Thanks for the good work
